aten/src/ATen/InferSize.h

```diff
@@ -37,31 +39,39 @@ inline void infer_size_impl(
     }
   }
 
-  if (TORCH_GUARD_SIZE_OBLIVIOUS(sym_eq(numel, newsize)) ||
-      (infer_dim && newsize > 0 && numel % newsize == 0)) {
-    if (infer_dim) {
-      // We have a degree of freedom here to select the dimension size; follow
-      // NumPy semantics and just bail.  However, a nice error message is needed
-      // because users often use `view` as a way to flatten & unflatten
-      // dimensions and will otherwise be confused why
-      //   empty_tensor.view( 0, 0)
-      // works yet
-      //   empty_tensor.view(-1, 0)
-      // doesn't.
-      TORCH_CHECK(
-          newsize != 0,
-          "cannot reshape tensor of 0 elements into shape ",
-          shape,
-          " because the unspecified dimension size -1 can be any "
-          "value and is ambiguous");
-      res[*infer_dim] = numel / newsize;
-    }
+  auto set_infer_dim = [&]() {
+    // We have a degree of freedom here to select the dimension size; follow
+    // NumPy semantics and just bail.  However, a nice error message is needed
+    // because users often use `view` as a way to flatten & unflatten
+    // dimensions and will otherwise be confused why
+    //   empty_tensor.view( 0, 0)
+    // works yet
+    //   empty_tensor.view(-1, 0)
+    // doesn't.
+    TORCH_CHECK(
+        newsize != 0,
+        "cannot reshape tensor of 0 elements into shape ",
+        shape,
+        " because the unspecified dimension size -1 can be any "
+        "value and is ambiguous");
+    res[*infer_dim] = numel / newsize;
+    return;
+  };
+
+  if (infer_dim && newsize > 0 && numel % newsize == 0) {
+    set_infer_dim();
     return;
   }
 
-  std::ostringstream ss;
-  ss << "shape '" << shape << "' is invalid for input of size " << numel;
-  throw std::runtime_error(ss.str());
+  TORCH_MAYBE_SYM_CHECK(
+      sym_eq(numel, newsize),
+      "shape '",
+      shape,
+      "' is invalid for input of size ",
+      numel);
+  if (infer_dim) {
+    set_infer_dim();
+  }
```

Comment:
The old code ALWAYS threw an exception here but the new code might not... intended?

Reply:
it did not always throw before there was a return at line 59 in the old code.

Reply:
the conditions are maintained if you try to decompose and flatten the checks